File: ecommerce_ds/ecommerce.py
# e_commerce_solution.py
# Solução de referência do sistema de E-commerce em Python.
# Objetivo didático: mostrar uma implementação clara a partir do esqueleto.
# Observações:
# - Para dinheiro, usamos decimal.Decimal (evita erros de ponto flutuante).
# - IDs são gerados com uuid.uuid4().
# - Estoque é checado/debitado somente na confirmação do pedido (fluxo mais realista).

import uuid                             # gera identificadores únicos (UUID)
from decimal import Decimal             # tipo apropriado para dinheiro
from typing import List, Dict, Optional # tipos para anotações estáticas


# =========================
# CLASSE PESSOA
# =========================
class Pessoa:
    """
    Representa uma pessoa com atributos básicos.
    Atributos: nome (str), cpf (str), email (str).
    """

    def __init__(self, nome: str, cpf: str, email: str) -> None:
        # Armazena os atributos informados
        self.nome = nome
        self.cpf = cpf
        self.email = email
        # Nota: validações (CPF, email) poderiam ser adicionadas aqui como extensão.

    def __repr__(self) -> str:
        # Representação útil para depuração e listagem
        return f"Pessoa(nome='{self.nome}', cpf='{self.cpf}', email='{self.email}')"


# =========================
# CLASSE CLIENTE
# =========================
class Cliente:
    """
    Representa um cliente do e-commerce.
    Atributos: pessoa (Pessoa), id_cliente (str).
    id_cliente é gerado automaticamente via UUID.
    """

    def __init__(self, pessoa: Pessoa) -> None:
        self.pessoa = pessoa
        self.id_cliente = str(uuid.uuid4())  # gera um identificador único

    def __repr__(self) -> str:
        return f"Cliente(id_cliente='{self.id_cliente}', pessoa={self.pessoa})"


# =========================
# CLASSE PRODUTO
# =========================
class Produto:
    """
    Representa um produto à venda.
    Atributos: nome (str), preco (Decimal), estoque (int).
    """

    def __init__(self, nome: str, preco: Decimal, estoque: int) -> None:
        # Guarda os atributos; conversão e validação simples
        self.nome = nome
        self.preco = Decimal(preco)          # garante que seja Decimal
        if self.preco < 0:
            raise ValueError("Preço não pode ser negativo.")
        if estoque < 0:
            raise ValueError("Estoque não pode ser negativo.")
        self.estoque = int(estoque)

    def __repr__(self) -> str:
        return f"Produto(nome='{self.nome}', preco={self.preco}, estoque={self.estoque})"


# =========================
# CLASSE ITEMPEDIDO
# =========================
class ItemPedido:
    """
    Representa um item dentro de um pedido: produto + quantidade.
    """

    def __init__(self, produto: Produto, quantidade: int) -> None:
        if quantidade <= 0:
            raise ValueError("Quantidade deve ser positiva.")
        self.produto = produto
        self.quantidade = int(quantidade)

    def calcular_total(self) -> Decimal:
        """
        Retorna o subtotal deste item: preco_unitario * quantidade.
        """
        return self.produto.preco * Decimal(self.quantidade)

    def __repr__(self) -> str:
        return f"ItemPedido(produto='{self.produto.nome}', quantidade={self.quantidade}, subtotal={self.calcular_total()})"


# =========================
# CLASSE PEDIDO
# =========================
class Pedido:
    """
    Representa um pedido realizado por um cliente.
    Atributos: cliente (Cliente), itens (List[ItemPedido]).
    Fluxo recomendado:
      - Criar pedido
      - Adicionar itens
      - Confirmar (valida e debita estoques)
    """

    def __init__(self, cliente: Cliente) -> None:
        self.cliente = cliente              # referência ao cliente que faz o pedido
        self.itens: List[ItemPedido] = []   # lista de itens inicialmente vazia
        self.id_pedido: str = str(uuid.uuid4())  # ID do pedido (opcional, mas prático)
        self.confirmado: bool = False       # status do pedido

    def adicionar_item(self, produto: Produto, quantidade: int) -> None:
        """
        Cria um ItemPedido e adiciona à lista de itens.
        """
        if self.confirmado:
            # Impede adicionar itens após confirmação (regra comum)
            raise RuntimeError("Não é possível adicionar itens em um pedido já confirmado.")
        self.itens.append(ItemPedido(produto, quantidade))

    def calcular_total(self) -> Decimal:
        """
        Soma os subtotais de todos os itens do pedido.
        """
        total = Decimal("0.00")
        for it in self.itens:
            total += it.calcular_total()
        return total

    def confirmar(self) -> bool:
        """
        Valida se há estoque suficiente para TODOS os itens.
        Se houver, debita os estoques e marca o pedido como confirmado.
        Retorna True em caso de sucesso; False se faltar estoque para algum produto.
        """
        if self.confirmado:
            return True  # já está confirmado

        # 1) Agrega as quantidades por produto para validação correta (mesmo produto repetido)
        demanda: Dict[Produto, int] = {}
        for it in self.itens:
            demanda[it.produto] = demanda.get(it.produto, 0) + it.quantidade

        # 2) Verifica se todos os produtos possuem estoque suficiente
        for produto, qtd_total in demanda.items():
            if produto.estoque < qtd_total:
                return False  # se faltar estoque para qualquer produto, não confirma

        # 3) Debita o estoque de todos os produtos (agora com segurança)
        for produto, qtd_total in demanda.items():
            produto.estoque -= qtd_total

        self.confirmado = True
        return True

    def __repr__(self) -> str:
        return f"Pedido(id='{self.id_pedido}', cliente='{self.cliente.pessoa.nome}', itens={len(self.itens)}, total={self.calcular_total()}, confirmado={self.confirmado})"


# =========================
# CLASSE MENU (CLI)
# =========================
class Menu:
    """
    Gerencia o e-commerce em memória.
    Estruturas:
      - clientes: Dict[str, Cliente]  (chave = CPF)
      - produtos: Dict[str, Produto]  (chave = ID do produto, gerado via UUID)
      - pedidos:  List[Pedido]
    """

    def __init__(self) -> None:
        # Dicionário de clientes indexado por CPF (acesso direto ao buscar cliente)
        self.clientes: Dict[str, Cliente] = {}
        # Dicionário de produtos indexado por ID (string UUID)
        self.produtos: Dict[str, Produto] = {}
        # Lista de pedidos confirmados
        self.pedidos: List[Pedido] = {}

        # Corrigindo: pedidos deve ser lista, não dicionário
        self.pedidos = []

    # -----------------------
    # CADASTRAR CLIENTE
    # -----------------------
    def cadastrar_cliente(self, nome: str, cpf: str, email: str) -> None:
        """
        Cria Pessoa -> Cliente e salva em self.clientes por CPF.
        Se já existir um cliente com o CPF, apenas substitui/atualiza.
        """
        pessoa = Pessoa(nome, cpf, email)
        cliente = Cliente(pessoa)
        self.clientes[cpf] = cliente
        print(f"[OK] Cliente cadastrado: {cliente}")

    # -----------------------
    # CADASTRAR PRODUTO
    # -----------------------
    def cadastrar_produto(self, nome: str, preco: Decimal, estoque: int) -> None:
        """
        Cria Produto e armazena em self.produtos sob um ID (UUID).
        """
        produto = Produto(nome, Decimal(preco), estoque)
        id_produto = str(uuid.uuid4())
        self.produtos[id_produto] = produto
        print(f"[OK] Produto cadastrado: {id_produto} | {produto.nome} | R${produto.preco} | estoque={produto.estoque}")

    # -----------------------
    # CRIAR PEDIDO
    # -----------------------
    def criar_pedido(self, cpf: str) -> None:
        """
        Fluxo:
          - Busca cliente por CPF.
          - Permite adicionar itens (loop).
          - Exibe total.
          - Pergunta se confirma; se sim, valida e debita estoque.
        """
        cliente = self.clientes.get(cpf)
        if not cliente:
            print("[ERRO] Cliente não encontrado para o CPF informado.")
            return

        pedido = Pedido(cliente)  # cria um pedido em aberto
        print(f"== Novo pedido para: {cliente.pessoa.nome}")

        # Laço de inclusão de itens
        while True:
            self._listar_produtos_sintetico()
            prod_id = input("ID do produto (ENTER para encerrar): ").strip()
            if prod_id == "":
                break  # encerra inclusão

            produto = self.produtos.get(prod_id)
            if not produto:
                print("[ERRO] Produto não encontrado.")
                continue

            # Lê quantidade
            try:
                qtd = int(input("Quantidade: ").strip())
                if qtd <= 0:
                    print("[ERRO] Quantidade deve ser positiva.")
                    continue
            except ValueError:
                print("[ERRO] Quantidade inválida.")
                continue

            # Adiciona item ao pedido (estoque será checado na confirmação)
            try:
                pedido.adicionar_item(produto, qtd)
                print(f"[OK] Item adicionado: {produto.nome} x {qtd}")
            except Exception as e:
                print(f"[ERRO] Falha ao adicionar item: {e}")

        # Exibe resumo do pedido e total
        if not pedido.itens:
            print("[AVISO] Pedido vazio. Nada a confirmar.")
            return

        print("\nItens do pedido:")
        for it in pedido.itens:
            print(f" - {it}")

        print(f"Total do pedido: R${pedido.calcular_total()}")

        # Confirmação do pedido: valida estoque e debita
        confirma = input("Confirmar pedido? (s/n): ").strip().lower()
        if confirma == "s":
            if pedido.confirmar():
                self.pedidos.append(pedido)  # guarda apenas pedidos confirmados
                print(f"[OK] Pedido confirmado! ID: {pedido.id_pedido}")
            else:
                print("[ERRO] Estoque insuficiente para um ou mais itens. Pedido não confirmado.")
        else:
            print("[INFO] Pedido cancelado pelo usuário.")

    # -----------------------
    # LISTAR PEDIDOS
    # -----------------------
    def listar_pedidos(self) -> None:
        """
        Lista pedidos confirmados com um resumo.
        """
        if not self.pedidos:
            print("[INFO] Não há pedidos confirmados.")
            return

        print("== Pedidos confirmados ==")
        for p in self.pedidos:
            print(f"- ID: {p.id_pedido} | Cliente: {p.cliente.pessoa.nome} | Itens: {len(p.itens)} | Total: R${p.calcular_total()}")

    # -----------------------
    # MENU (CLI)
    # -----------------------
    def exibir_menu(self) -> None:
        """
        Menu de linha de comando simples para interação manual.
        """
        while True:
            print("\n=== MENU ===")
            print("1) Cadastrar cliente")
            print("2) Cadastrar produto")
            print("3) Criar pedido")
            print("4) Listar pedidos")
            print("0) Sair")
            op = input("Opção: ").strip()

            try:
                if op == "1":
                    nome = input("Nome: ").strip()
                    cpf = input("CPF: ").strip()
                    email = input("Email: ").strip()
                    self.cadastrar_cliente(nome, cpf, email)

                elif op == "2":
                    nome = input("Nome do produto: ").strip()
                    preco_str = input("Preço (ex: 199.90): ").strip().replace(",", ".")
                    estoque_str = input("Estoque (inteiro): ").strip()
                    preco = Decimal(preco_str)
                    estoque = int(estoque_str)
                    self.cadastrar_produto(nome, preco, estoque)

                elif op == "3":
                    cpf = input("CPF do cliente: ").strip()
                    self.criar_pedido(cpf)

                elif op == "4":
                    self.listar_pedidos()

                elif op == "0":
                    print("Encerrando...")
                    return

                else:
                    print("[ERRO] Opção inválida.")

            except Exception as e:
                # Captura genérica para manter a CLI ativa mesmo em caso de erro
                print(f"[ERRO] {e}")

    # -----------------------
    # Utilitário: lista produtos
    # -----------------------
    def _listar_produtos_sintetico(self) -> None:
        """
        Exibe a lista de produtos com seus IDs (chave do dicionário), nome, preço e estoque.
        Facilita a seleção por ID na criação do pedido.
        """
        print("\n== Produtos cadastrados ==")
        if not self.produtos:
            print("(vazio)")
            return

        for prod_id, produto in self.produtos.items():
            print(f"{prod_id} | {produto.nome} | R${produto.preco} | estoque={produto.estoque}")


# =========================
# PONTO DE ENTRADA
# =========================
if __name__ == "__main__":
    # Instancia o Menu e inicia a CLI.
    menu = Menu()
    menu.exibir_menu()

File: ecommerce_ds/Ecommerce.java
// Main.java
// Sistema de E-commerce (versão Java) baseado no enunciado do PDF e no template Python.
// Comentários detalhados foram incluídos para facilitar o uso didático.

import java.math.BigDecimal;           // Representação correta de valores monetários
import java.util.*;                    // List, Map, Scanner, UUID etc.

////////////////////////////////////////////////////////////////////////////////
// Classe Pessoa: representa dados básicos (nome, cpf, email) usados por Cliente
////////////////////////////////////////////////////////////////////////////////
class Pessoa {
    // Atributos privados para manter encapsulamento
    private final String nome;   // nome completo da pessoa
    private final String cpf;    // documento no formato texto (não calcularemos máscara aqui)
    private final String email;  // email de contato

    // Construtor: recebe os três campos obrigatórios
    public Pessoa(String nome, String cpf, String email) {
        this.nome  = Objects.requireNonNull(nome,  "nome obrigatório");
        this.cpf   = Objects.requireNonNull(cpf,   "cpf obrigatório");
        this.email = Objects.requireNonNull(email, "email obrigatório");
    }

    // Getters: acesso somente leitura
    public String getNome()  { return nome; }
    public String getCpf()   { return cpf; }
    public String getEmail() { return email; }

    // toString: útil para depuração e listagens
    @Override
    public String toString() {
        return "Pessoa{nome='" + nome + "', cpf='" + cpf + "', email='" + email + "'}";
    }
}

////////////////////////////////////////////////////////////////////////////////
// Classe Cliente: associa uma Pessoa e um identificador único (UUID)
////////////////////////////////////////////////////////////////////////////////
class Cliente {
    private final Pessoa pessoa;        // composição: Cliente contém uma Pessoa
    private final String idCliente;     // identificador único como String (UUID)

    // Construtor: gera um UUID automaticamente
    public Cliente(Pessoa pessoa) {
        this.pessoa = Objects.requireNonNull(pessoa, "pessoa obrigatória");
        this.idCliente = UUID.randomUUID().toString(); // equivalente a uuid.uuid4() no Python
    }

    public Pessoa getPessoa()      { return pessoa; }
    public String getIdCliente()   { return idCliente; }

    @Override
    public String toString() {
        return "Cliente{id='" + idCliente + "', " + pessoa + "}";
    }
}

////////////////////////////////////////////////////////////////////////////////
// Classe Produto: representa item vendável com nome, preço (BigDecimal) e estoque
////////////////////////////////////////////////////////////////////////////////
class Produto {
    private final String idProduto;    // identificador único do produto
    private String nome;               // nome do produto
    private BigDecimal preco;          // preço unitário (BigDecimal evita problemas de ponto flutuante)
    private int estoque;               // quantidade disponível

    // Construtor: inicializa todos os campos
    public Produto(String nome, BigDecimal preco, int estoque) {
        this.idProduto = UUID.randomUUID().toString();   // gera id único
        this.nome = Objects.requireNonNull(nome, "nome obrigatório");
        this.preco = Objects.requireNonNull(preco, "preço obrigatório");
        if (estoque < 0) throw new IllegalArgumentException("estoque não pode ser negativo");
        this.estoque = estoque;
    }

    // Getters e setters mínimos (apenas onde faz sentido mudar)
    public String getIdProduto()  { return idProduto; }
    public String getNome()       { return nome; }
    public BigDecimal getPreco()  { return preco; }
    public int getEstoque()       { return estoque; }

    public void setNome(String nome)            { this.nome = Objects.requireNonNull(nome); }
    public void setPreco(BigDecimal novoPreco)  { this.preco = Objects.requireNonNull(novoPreco); }

    // Métodos de estoque: debitar/creditar com validação simples
    public boolean podeAtender(int quantidade) {
        return quantidade > 0 && estoque >= quantidade;
    }
    public void debitarEstoque(int quantidade) {
        if (!podeAtender(quantidade)) throw new IllegalStateException("estoque insuficiente");
        this.estoque -= quantidade;
    }
    public void creditarEstoque(int quantidade) {
        if (quantidade < 0) throw new IllegalArgumentException("quantidade inválida");
        this.estoque += quantidade;
    }

    @Override
    public String toString() {
        return "Produto{id='" + idProduto + "', nome='" + nome + "', preco=" + preco + ", estoque=" + estoque + "}";
    }
}

////////////////////////////////////////////////////////////////////////////////
// Classe ItemPedido: liga um Produto a uma quantidade e calcula subtotal
////////////////////////////////////////////////////////////////////////////////
class ItemPedido {
    private final Produto produto;  // referência ao produto escolhido
    private final int quantidade;   // quantidade solicitada

    public ItemPedido(Produto produto, int quantidade) {
        if (quantidade <= 0) throw new IllegalArgumentException("quantidade deve ser > 0");
        this.produto = Objects.requireNonNull(produto, "produto obrigatório");
        this.quantidade = quantidade;
    }

    public Produto getProduto()   { return produto; }
    public int getQuantidade()    { return quantidade; }

    // calcular_total: preço unitário * quantidade
    public BigDecimal calcularTotal() {
        return produto.getPreco().multiply(BigDecimal.valueOf(quantidade));
    }

    @Override
    public String toString() {
        return "ItemPedido{produto=" + produto.getNome() + ", qtd=" + quantidade + ", subtotal=" + calcularTotal() + "}";
    }
}

////////////////////////////////////////////////////////////////////////////////
// Enum StatusPedido: ajuda a tornar explícito o ciclo de vida do pedido
////////////////////////////////////////////////////////////////////////////////
enum StatusPedido { ABERTO, CONFIRMADO, CANCELADO }

////////////////////////////////////////////////////////////////////////////////
// Classe Pedido: contém o Cliente e uma lista de itens; confirma e calcula total
////////////////////////////////////////////////////////////////////////////////
class Pedido {
    private final String idPedido;         // identificador único do pedido
    private final Cliente cliente;         // quem fez o pedido
    private final List<ItemPedido> itens;  // itens (produto + quantidade)
    private StatusPedido status;           // estado atual (ABERTO/CONFIRMADO/CANCELADO)

    // Construtor: inicia a lista de itens vazia e status ABERTO
    public Pedido(Cliente cliente) {
        this.idPedido = UUID.randomUUID().toString();
        this.cliente = Objects.requireNonNull(cliente, "cliente obrigatório");
        this.itens = new ArrayList<>();
        this.status = StatusPedido.ABERTO;
    }

    public String getIdPedido()         { return idPedido; }
    public Cliente getCliente()         { return cliente; }
    public List<ItemPedido> getItens()  { return Collections.unmodifiableList(itens); }
    public StatusPedido getStatus()     { return status; }

    // adicionar_item: apenas adiciona enquanto o pedido está ABERTO (não debita estoque aqui)
    public void adicionarItem(Produto produto, int quantidade) {
        if (status != StatusPedido.ABERTO) throw new IllegalStateException("pedido não está ABERTO");
        itens.add(new ItemPedido(produto, quantidade));
    }

    // calcular_total: soma dos subtotais
    public BigDecimal calcularTotal() {
        BigDecimal total = BigDecimal.ZERO;
        for (ItemPedido it : itens) {
            total = total.add(it.calcularTotal());
        }
        return total;
    }

    // confirmar: valida estoques de todos os itens e então debita de cada produto
    public boolean confirmar() {
        if (status != StatusPedido.ABERTO) return false;         // só confirma uma vez
        // 1) valida se todos os produtos têm estoque suficiente
        for (ItemPedido it : itens) {
            if (!it.getProduto().podeAtender(it.getQuantidade())) {
                return false; // se algum falhar, aborta sem debitar nada
            }
        }
        // 2) debita dos estoques (agora com segurança)
        for (ItemPedido it : itens) {
            it.getProduto().debitarEstoque(it.getQuantidade());
        }
        status = StatusPedido.CONFIRMADO;                         // atualiza status
        return true;
    }

    // cancelar: apenas troca o status se ainda estiver ABERTO (não há débito para reverter)
    public void cancelar() {
        if (status == StatusPedido.ABERTO) status = StatusPedido.CANCELADO;
    }

    @Override
    public String toString() {
        return "Pedido{id='" + idPedido + "', cliente=" + cliente.getPessoa().getNome()
               + ", itens=" + itens.size() + ", total=" + calcularTotal() + ", status=" + status + "}";
    }
}

////////////////////////////////////////////////////////////////////////////////
// Classe Menu: gerencia cadastros e a interação via linha de comando
////////////////////////////////////////////////////////////////////////////////
class Menu {
    // Estruturas de dados do “sistema” (em memória):
    // - clientes por CPF (Map para acesso rápido)
    // - produtos por ID (Map) e lista de pedidos
    private final Map<String, Cliente> clientesPorCpf = new HashMap<>();
    private final Map<String, Produto> produtosPorId  = new HashMap<>();
    private final List<Pedido> pedidos                = new ArrayList<>();
    private final Scanner in = new Scanner(System.in); // leitura de entradas do usuário

    // cadastrar_cliente: cria Pessoa e Cliente e guarda por CPF
    public void cadastrarCliente(String nome, String cpf, String email) {
        Pessoa p = new Pessoa(nome, cpf, email);
        Cliente c = new Cliente(p);
        clientesPorCpf.put(p.getCpf(), c);
        System.out.println("Cliente cadastrado: " + c);
    }

    // cadastrar_produto: cria Produto e adiciona ao Map por ID
    public void cadastrarProduto(String nome, BigDecimal preco, int estoque) {
        Produto pr = new Produto(nome, preco, estoque);
        produtosPorId.put(pr.getIdProduto(), pr);
        System.out.println("Produto cadastrado: " + pr);
    }

    // criar_pedido: recebe CPF, permite adicionar itens e, ao final, confirmar (debitando estoque)
    public void criarPedido(String cpf) {
        Cliente c = clientesPorCpf.get(cpf);
        if (c == null) {
            System.out.println("Cliente não encontrado para CPF: " + cpf);
            return;
        }
        Pedido pedido = new Pedido(c);
        System.out.println("== Novo pedido para " + c.getPessoa().getNome());

        // Laço de inclusão de itens
        while (true) {
            listarProdutosSintetico(); // mostra ids e estoques atuais
            System.out.print("Digite o ID do produto (ou ENTER para encerrar a inclusão): ");
            String id = in.nextLine().trim();
            if (id.isEmpty()) break;

            Produto pr = produtosPorId.get(id);
            if (pr == null) {
                System.out.println("Produto não encontrado.");
                continue;
            }

            System.out.print("Quantidade: ");
            int qtd;
            try {
                qtd = Integer.parseInt(in.nextLine().trim());
            } catch (NumberFormatException ex) {
                System.out.println("Quantidade inválida.");
                continue;
            }

            // Adiciona no pedido (a validação de estoque será feita na confirmação)
            try {
                pedido.adicionarItem(pr, qtd);
                System.out.println("Item adicionado: " + pr.getNome() + " x " + qtd);
            } catch (Exception e) {
                System.out.println("Falha ao adicionar item: " + e.getMessage());
            }
        }

        // Mostra resumo e total
        System.out.println("Itens no pedido:");
        for (ItemPedido it : pedido.getItens()) {
            System.out.println(" - " + it);
        }
        System.out.println("Total: " + pedido.calcularTotal());

        // Confirma e debita estoque conforme o enunciado do PDF
        System.out.print("Confirmar pedido? (s/n): ");
        String conf = in.nextLine().trim().toLowerCase();
        if (conf.equals("s")) {
            boolean ok = pedido.confirmar();
            if (ok) {
                pedidos.add(pedido); // só adiciona à lista se confirmado
                System.out.println("Pedido confirmado! ID: " + pedido.getIdPedido());
            } else {
                System.out.println("Não foi possível confirmar: estoque insuficiente em algum item.");
                pedido.cancelar();
            }
        } else {
            pedido.cancelar();
            System.out.println("Pedido cancelado.");
        }
    }

    // listar_pedidos: percorre a lista mostrando um resumo
    public void listarPedidos() {
        if (pedidos.isEmpty()) {
            System.out.println("Não há pedidos confirmados.");
            return;
        }
        for (Pedido p : pedidos) {
            System.out.println(p);
        }
    }

    // Utilitário: mostra produtos resumidos (ID, nome, preço, estoque)
    private void listarProdutosSintetico() {
        System.out.println("== Produtos cadastrados ==");
        if (produtosPorId.isEmpty()) {
            System.out.println("Nenhum produto cadastrado.");
            return;
        }
        for (Produto pr : produtosPorId.values()) {
            System.out.println(pr.getIdProduto() + " | " + pr.getNome()
                    + " | R$" + pr.getPreco() + " | estoque=" + pr.getEstoque());
        }
    }

    // exibir_menu: loop principal de interação com o usuário
    public void exibirMenu() {
        while (true) {
            System.out.println("\n=== MENU ===");
            System.out.println("1) Cadastrar cliente");
            System.out.println("2) Cadastrar produto");
            System.out.println("3) Criar pedido");
            System.out.println("4) Listar pedidos");
            System.out.println("0) Sair");
            System.out.print("Opção: ");

            String op = in.nextLine().trim();
            try {
                switch (op) {
                    case "1":
                        System.out.print("Nome: ");  String nome = in.nextLine().trim();
                        System.out.print("CPF: ");   String cpf  = in.nextLine().trim();
                        System.out.print("Email: "); String email= in.nextLine().trim();
                        cadastrarCliente(nome, cpf, email);
                        break;
                    case "2":
                        System.out.print("Nome: "); String nomeP = in.nextLine().trim();
                        System.out.print("Preço (ex: 199.90): ");
                        BigDecimal preco = new BigDecimal(in.nextLine().trim());
                        System.out.print("Estoque: ");
                        int estoque = Integer.parseInt(in.nextLine().trim());
                        cadastrarProduto(nomeP, preco, estoque);
                        break;
                    case "3":
                        System.out.print("CPF do cliente: ");
                        String cpfCli = in.nextLine().trim();
                        criarPedido(cpfCli);
                        break;
                    case "4":
                        listarPedidos();
                        break;
                    case "0":
                        System.out.println("Encerrando...");
                        return; // sai do método e encerra o programa
                    default:
                        System.out.println("Opção inválida.");
                }
            } catch (Exception e) {
                // Captura genérica para manter o programa vivo em caso de erro de entrada
                System.out.println("Erro: " + e.getMessage());
            }
        }
    }
}

////////////////////////////////////////////////////////////////////////////////
// Classe pública com método main: ponto de entrada do programa
////////////////////////////////////////////////////////////////////////////////
public class Ecommerce {
    public static void main(String[] args) {
        // Cria o menu e inicia o loop de interação
        Menu menu = new Menu();
        menu.exibirMenu();
    }
}
